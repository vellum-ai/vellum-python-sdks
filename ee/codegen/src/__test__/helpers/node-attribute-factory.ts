import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
