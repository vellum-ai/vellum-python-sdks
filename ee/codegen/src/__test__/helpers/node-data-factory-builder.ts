import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
