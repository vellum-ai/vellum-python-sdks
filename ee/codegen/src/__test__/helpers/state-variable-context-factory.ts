import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
