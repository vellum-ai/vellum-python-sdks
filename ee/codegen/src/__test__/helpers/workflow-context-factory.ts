import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
