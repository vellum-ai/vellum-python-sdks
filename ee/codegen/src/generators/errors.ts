export type CodegenErrorCode =
  | "PROJECT_SERIALIZATION_ERROR"
  | "WORKFLOW_GENERATION_ERROR"
  | "WORKFLOW_INPUT_GENERATION_ERROR"
  | "NODE_DEFINITION_GENERATION_ERROR"
  | "NODE_ATTRIBUTE_GENERATION_ERROR"
  | "NODE_PORT_GENERATION_ERROR"
  | "NODE_NOT_FOUND_ERROR"
  | "NODE_OUTPUT_NOT_FOUND_ERROR"
  | "NODE_INPUT_NOT_FOUND_ERROR"
  | "UNSUPPORTED_SANDBOX_INPUT_ERROR"
  | "ENTITY_NOT_FOUND_ERROR"
  | "POST_PROCESSING_ERROR"
  | "VALUE_GENERATION_ERROR";

export abstract class BaseCodegenError extends Error {
  abstract code: CodegenErrorCode;
}

/**
 * An error that raises when deserializing the request to codegen
 * into a valid Workflow project.
 */
export class ProjectSerializationError extends BaseCodegenError {
  code = "PROJECT_SERIALIZATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow definition fails to
 * generate.
 */
export class WorkflowGenerationError extends BaseCodegenError {
  code = "WORKFLOW_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow Inpus fail to
 * generate.
 */
export class WorkflowInputGenerationError extends BaseCodegenError {
  code = "WORKFLOW_INPUT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Node definition fails to generate
 */
export class NodeDefinitionGenerationError extends BaseCodegenError {
  code = "NODE_DEFINITION_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node attribute fails.
 */
export class NodeAttributeGenerationError extends BaseCodegenError {
  code = "NODE_ATTRIBUTE_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node port fails.
 */
export class NodePortGenerationError extends BaseCodegenError {
  code = "NODE_PORT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when a node is not found.
 */
export class NodeNotFoundError extends BaseCodegenError {
  code = "NODE_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a vellum entity is not found.
 */
export class EntityNotFoundError extends BaseCodegenError {
  code = "ENTITY_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a sandbox input is not supported.
 */
export class UnsupportedSandboxInputError extends BaseCodegenError {
  code = "UNSUPPORTED_SANDBOX_INPUT_ERROR" as const;
}

/**
 * An error that raises when a node output is not found.
 */
export class NodeOutputNotFoundError extends BaseCodegenError {
  code = "NODE_OUTPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a node input is not found.
 */
export class NodeInputNotFoundError extends BaseCodegenError {
  code = "NODE_INPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when the code fails to post process
 */
export class PostProcessingError extends BaseCodegenError {
  code = "POST_PROCESSING_ERROR" as const;
}

/**
 * An error that raises when a value fails to generate
 */
export class ValueGenerationError extends BaseCodegenError {
  code = "VALUE_GENERATION_ERROR" as const;
}
