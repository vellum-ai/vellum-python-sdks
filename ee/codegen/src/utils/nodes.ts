import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
