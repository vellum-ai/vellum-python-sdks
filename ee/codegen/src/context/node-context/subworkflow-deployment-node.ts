import { VellumVariableType, WorkflowDeploymentRelease } from "vellum-ai/api";
import { ReleaseReviews as WorkflowReleaseClient } from "vellum-ai/api/resources/releaseReviews/client/Client";

import { BaseNodeContext } from "./base";

import { PortContext } from "src/context/port-context";
import { NodeDefinitionGenerationError } from "src/generators/errors";
import { SubworkflowNode as SubworkflowNodeType } from "src/types/vellum";
import { toValidPythonIdentifier } from "src/utils/casing";
import { isVellumErrorWithDetail } from "src/utils/nodes";

export class SubworkflowDeploymentNodeContext extends BaseNodeContext<SubworkflowNodeType> {
  baseNodeClassName = "SubworkflowDeploymentNode";
  baseNodeDisplayClassName = "BaseSubworkflowDeploymentNodeDisplay";

  public workflowDeploymentRelease: WorkflowDeploymentRelease | null = null;

  getNodeOutputNamesById(): Record<string, string> {
    if (!this.workflowDeploymentRelease) {
      return {};
    }

    const errorOutputId = this.getErrorOutputId();
    return this.workflowDeploymentRelease.workflowVersion.outputVariables.reduce<
      Record<string, string>
    >(
      (acc, output) => {
        acc[output.id] = toValidPythonIdentifier(output.key, "output");
        return acc;
      },
      { ...(errorOutputId ? { [errorOutputId]: "error" } : {}) }
    );
  }

  getNodeOutputTypesById(): Record<string, VellumVariableType> {
    if (!this.workflowDeploymentRelease) {
      return {};
    }

    return Object.fromEntries(
      this.workflowDeploymentRelease.workflowVersion.outputVariables.map(
        (variable) => [variable.id, variable.type]
      )
    );
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }

  async buildProperties(): Promise<void> {
    if (this.nodeData.data.variant !== "DEPLOYMENT") {
      return;
    }

    try {
      this.workflowDeploymentRelease = await new WorkflowReleaseClient({
        apiKey: this.workflowContext.vellumApiKey,
        environment: this.workflowContext.vellumApiEnvironment,
      }).retrieveWorkflowDeploymentRelease(
        this.nodeData.data.workflowDeploymentId,
        this.nodeData.data.releaseTag
      );
    } catch (error) {
      if (isVellumErrorWithDetail(error)) {
        throw new NodeDefinitionGenerationError(error.body.detail, "WARNING");
      }

      throw error;
    }
  }
}
