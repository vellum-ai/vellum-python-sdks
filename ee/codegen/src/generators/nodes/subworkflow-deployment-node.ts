import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { SubworkflowDeploymentNodeContext } from "src/context/node-context/subworkflow-deployment-node";
import {
  NodeAttributeGenerationError,
  NodeDefinitionGenerationError,
} from "src/generators/errors";
import { BaseNode } from "src/generators/nodes/bases/base";
import { codegen } from "src/index";
import { SubworkflowNode as SubworkflowNodeType } from "src/types/vellum";

const INPUTS_PREFIX = "subworkflow_inputs";

export class SubworkflowDeploymentNode extends BaseNode<
  SubworkflowNodeType,
  SubworkflowDeploymentNodeContext
> {
  protected getNodeAttributeNameByNodeInputKey(nodeInputKey: string): string {
    return `${INPUTS_PREFIX}.${nodeInputKey}`;
  }

  getNodeClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    if (this.nodeData.data.variant !== "DEPLOYMENT") {
      throw new NodeDefinitionGenerationError(
        `SubworkflowDeploymentNode only supports DEPLOYMENT variant. Received ${this.nodeData.data.variant}`
      );
    }

    if (!this.nodeContext.workflowDeploymentRelease) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          `Failed to generate attribute: ${this.nodeData.data.label}.deployment`
        )
      );
    } else {
      statements.push(
        python.field({
          name: "deployment",
          initializer: python.TypeInstantiation.str(
            this.nodeContext.workflowDeploymentRelease.deployment.name
          ),
        })
      );
    }

    statements.push(
      python.field({
        name: "release_tag",
        initializer: python.TypeInstantiation.str(
          this.nodeData.data.releaseTag
        ),
      })
    );

    statements.push(
      python.field({
        name: INPUTS_PREFIX,
        initializer: python.TypeInstantiation.dict(
          Array.from(this.nodeInputsByKey.entries()).map(([key, value]) => ({
            key: python.TypeInstantiation.str(key),
            value: value,
          })),
          {
            endWithComma: true,
          }
        ),
      })
    );

    const outputsClass = this.generateOutputsClass();
    if (outputsClass) {
      statements.push(outputsClass);
    }

    return statements;
  }

  private generateOutputsClass(): python.Class | null {
    if (!this.nodeContext.workflowDeploymentRelease) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          `Failed to generate ${this.nodeData.data.label}.Outputs class`
        )
      );
      return null;
    }

    const nodeBaseClassRef = this.getNodeBaseClass();
    const outputsClass = python.class_({
      name: OUTPUTS_CLASS_NAME,
      extends_: [
        python.reference({
          name: nodeBaseClassRef.name,
          modulePath: nodeBaseClassRef.modulePath,
          alias: nodeBaseClassRef.alias,
          attribute: [OUTPUTS_CLASS_NAME],
        }),
      ],
    });

    this.nodeContext.workflowDeploymentRelease.workflowVersion.outputVariables.forEach(
      (output) => {
        const outputName = this.nodeContext.getNodeOutputNameById(output.id);
        if (outputName) {
          outputsClass.add(
            codegen.vellumVariable({
              variable: {
                name: outputName,
                type: output.type,
                id: output.id,
                required: output.required,
              },
              defaultRequired: true,
            })
          );
        }
      }
    );

    return outputsClass;
  }

  getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "label",
        initializer: python.TypeInstantiation.str(this.nodeData.data.label),
      })
    );

    statements.push(
      python.field({
        name: "node_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.id),
      })
    );

    statements.push(
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    return statements;
  }

  protected getOutputDisplay(): python.Field | undefined {
    if (!this.nodeContext.workflowDeploymentRelease) {
      this.workflowContext.addError(
        new NodeDefinitionGenerationError(
          `Failed to generate \`output_display\` for ${this.nodeData.data.label}`,
          "WARNING"
        )
      );
    }

    const outputVariables =
      this.nodeContext.workflowDeploymentRelease?.workflowVersion
        .outputVariables ?? [];

    return python.field({
      name: "output_display",
      initializer: python.TypeInstantiation.dict(
        outputVariables.map((output) => {
          const outputName = this.nodeContext.getNodeOutputNameById(output.id);
          if (!outputName) {
            throw new NodeAttributeGenerationError(
              `Could not find output name for ${this.nodeContext.nodeClassName}.Outputs.${output.key} given output id ${output.id}`
            );
          }
          return {
            key: python.reference({
              name: this.nodeContext.nodeClassName,
              modulePath: this.nodeContext.nodeModulePath,
              attribute: [OUTPUTS_CLASS_NAME, outputName],
            }),
            value: python.instantiateClass({
              classReference: python.reference({
                name: "NodeOutputDisplay",
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .NODE_DISPLAY_TYPES_MODULE_PATH,
              }),
              arguments_: [
                python.methodArgument({
                  name: "id",
                  value: python.TypeInstantiation.uuid(output.id),
                }),
                python.methodArgument({
                  name: "name",
                  value: python.TypeInstantiation.str(output.key),
                }),
              ],
            }),
          };
        })
      ),
    });
  }

  protected getErrorOutputId(): string | undefined {
    return this.nodeData.data.errorOutputId;
  }
}
