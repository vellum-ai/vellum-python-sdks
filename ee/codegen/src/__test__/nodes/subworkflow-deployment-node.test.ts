import { Writer } from "@fern-api/python-ast/core/Writer";
import { WorkflowDeploymentHistoryItem } from "vellum-ai/api";
import { WorkflowDeployments as WorkflowDeploymentsClient } from "vellum-ai/api/resources/workflowDeployments/client/Client";
import { VellumError } from "vellum-ai/errors";
import { beforeEach, vi } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { subworkflowDeploymentNodeDataFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { SubworkflowDeploymentNodeContext } from "src/context/node-context/subworkflow-deployment-node";
import { NodeDefinitionGenerationError } from "src/generators/errors";
import { SubworkflowDeploymentNode } from "src/generators/nodes/subworkflow-deployment-node";

describe("SubworkflowDeploymentNode", () => {
  let workflowContext: WorkflowContext;
  let node: SubworkflowDeploymentNode;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("basic", () => {
    beforeEach(async () => {
      vi.spyOn(
        WorkflowDeploymentsClient.prototype,
        "workflowDeploymentHistoryItemRetrieve"
      ).mockResolvedValue({
        name: "test-deployment",
        outputVariables: [
          { id: "1", key: "output-1", type: "STRING" },
          { id: "2", key: "output-2", type: "NUMBER" },
        ],
      } as unknown as WorkflowDeploymentHistoryItem);

      const nodeData = subworkflowDeploymentNodeDataFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as SubworkflowDeploymentNodeContext;

      node = new SubworkflowDeploymentNode({
        workflowContext,
        nodeContext,
      });
    });

    it(`getNodeFile`, async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it(`getNodeDisplayFile`, async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("failure", () => {
    it(`should throw an error we can handle if the workflow deployment history item is not found`, async () => {
      vi.spyOn(
        WorkflowDeploymentsClient.prototype,
        "workflowDeploymentHistoryItemRetrieve"
      ).mockRejectedValue(
        new VellumError({
          body: {
            detail: "No Workflow Deployment found.",
          },
        })
      );

      const nodeData = subworkflowDeploymentNodeDataFactory();

      await expect(
        createNodeContext({
          workflowContext,
          nodeData,
        })
      ).rejects.toThrow(
        new NodeDefinitionGenerationError("No Workflow Deployment found.")
      );
    });

    it(`should generate subworkflow deployment nodes as much as possible for non strict workflow contexts`, async () => {
      const workflowContext = workflowContextFactory({
        strict: false,
      });

      vi.spyOn(
        WorkflowDeploymentsClient.prototype,
        "workflowDeploymentHistoryItemRetrieve"
      ).mockRejectedValue(
        new VellumError({
          body: {
            detail: "No Workflow Deployment found.",
          },
        })
      );

      const nodeData = subworkflowDeploymentNodeDataFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as SubworkflowDeploymentNodeContext;

      node = new SubworkflowDeploymentNode({
        workflowContext,
        nodeContext,
      });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it(`should generate subworkflow deployment node display as much as possible for non strict workflow contexts`, async () => {
      const workflowContext = workflowContextFactory({
        strict: false,
      });

      vi.spyOn(
        WorkflowDeploymentsClient.prototype,
        "workflowDeploymentHistoryItemRetrieve"
      ).mockRejectedValue(
        new VellumError({
          body: {
            detail: "No Workflow Deployment found.",
          },
        })
      );

      const nodeData = subworkflowDeploymentNodeDataFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as SubworkflowDeploymentNodeContext;

      node = new SubworkflowDeploymentNode({
        workflowContext,
        nodeContext,
      });

      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
