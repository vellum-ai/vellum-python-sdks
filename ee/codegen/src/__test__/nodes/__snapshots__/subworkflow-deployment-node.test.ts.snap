// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubworkflowDeploymentNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import BaseSubworkflowDeploymentNodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)

from ...nodes.subworkflow_node import SubworkflowNode


class SubworkflowNodeDisplay(BaseSubworkflowDeploymentNodeDisplay[SubworkflowNode]):
    label = "Subworkflow Node"
    node_id = UUID("c8f2964c-09b8-44e0-a06d-606319fe5e2a")
    target_handle_id = UUID("f5e6bd33-527a-4ba6-8906-cd5e96a4321c")
    output_display = {
        SubworkflowNode.Outputs.output_1: NodeOutputDisplay(
            id=UUID("1"), name="output-1"
        ),
        SubworkflowNode.Outputs.output_2: NodeOutputDisplay(
            id=UUID("2"), name="output-2"
        ),
    }
    port_displays = {
        SubworkflowNode.Ports.default: PortDisplayOverrides(
            id=UUID("600efd51-8677-4ba3-a582-b298bebb2868")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2239.986322714681, y=484.74458968144046)
    )
"
`;

exports[`SubworkflowDeploymentNode > basic > getNodeFile 1`] = `
"from typing import Union

from vellum.workflows.nodes.displayable import SubworkflowDeploymentNode


class SubworkflowNode(SubworkflowDeploymentNode):
    deployment = "test-deployment"
    release_tag = "LATEST"
    subworkflow_inputs = {}

    class Outputs(SubworkflowDeploymentNode.Outputs):
        output_1: str
        output_2: Union[float, int]
"
`;

exports[`SubworkflowDeploymentNode > failure > should generate subworkflow deployment node display as much as possible for non strict workflow contexts 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import BaseSubworkflowDeploymentNodeDisplay
from vellum_ee.workflows.display.nodes.types import PortDisplayOverrides

from ...nodes.subworkflow_node import SubworkflowNode


class SubworkflowNodeDisplay(BaseSubworkflowDeploymentNodeDisplay[SubworkflowNode]):
    label = "Subworkflow Node"
    node_id = UUID("c8f2964c-09b8-44e0-a06d-606319fe5e2a")
    target_handle_id = UUID("f5e6bd33-527a-4ba6-8906-cd5e96a4321c")
    output_display = {}
    port_displays = {
        SubworkflowNode.Ports.default: PortDisplayOverrides(
            id=UUID("600efd51-8677-4ba3-a582-b298bebb2868")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2239.986322714681, y=484.74458968144046)
    )
"
`;

exports[`SubworkflowDeploymentNode > failure > should generate subworkflow deployment nodes as much as possible for non strict workflow contexts 1`] = `
"from vellum.workflows.nodes.displayable import SubworkflowDeploymentNode


class SubworkflowNode(SubworkflowDeploymentNode):
    release_tag = "LATEST"
    subworkflow_inputs = {}
"
`;
